packages/opencensus-core/src/trace/model/span.ts

```diff
@@ -13,60 +13,370 @@
+  constructor(parent?: Span) {
+    this.className = this.constructor.name;
+    this.id = randomSpanId();
+    this.spansLocal = [];
+    if (parent) {
+      this.root = parent.root;
+      this.parentSpan = parent;
+      this.activeTraceParams = this.root.activeTraceParams;
+    } else {
+      this.root = this;
+    }
+    this.logger = (this.root && this.root.logger) || this.logger;
   }
 
-  /** Gets trace id of span. */
+  /** Gets the trace ID. */
   get traceId(): string {
     return this.root.traceId;
   }
 
+  /** Gets the trace state */
   get traceState(): types.TraceState|undefined {
     return this.root.traceState;
   }
 
-  /** Starts the span instance. */
-  start() {
-    super.start();
-    this.logger.debug('starting span  %o', {
+  /**
+   * Gets the ID of the parent span.
+   * RootSpan doesn't have a parentSpan but it override this method.
+   */
+  get parentSpanId(): string {
+    if (!this.parentSpan) {
+      return 'no-parent';
+    }
+    return this.parentSpan.id;
+  }
+
+  /** Indicates if span was started. */
+  get started(): boolean {
+    return this.startedLocal;
+  }
+
+  /** Indicates if span was ended. */
+  get ended(): boolean {
+    return this.endedLocal;
+  }
+
+  /**
+   * Gives a timestamp that indicates the span's start time in RFC3339 UTC
+   * "Zulu" format.
+   */
+  get startTime(): Date {
+    if (!this.clock) {
+      this.logger.debug('calling startTime() on null clock');
+      return new Date();
+    }
+
+    return this.clock.startTime;
+  }
+
+  /** Recursively gets the descendant spans. */
+  allDescendants(): types.Span[] {
+    console.log('Calling allDescendants', this.name);
+    return this.spansLocal.reduce((acc: types.Span[], cur) => {
+      console.log('Name', cur.name, acc.length);
+      acc.push(cur);
+      const desc = cur.allDescendants();
+      acc = acc.concat(desc);
+      console.log(cur.name, acc.length);
+      return acc;
+    }, []);
+  }
+
+  /** The list of immediate child spans. */
+  get spans(): types.Span[] {
+    return this.spansLocal;
+  }
+
+  /** The number of direct children. */
+  get numberOfChildren(): number {
+    return this.spansLocal.length;
+  }
+
+  /**
+   * Gives a timestap that indicates the span's end time in RFC3339 UTC
+   * "Zulu" format.
+   */
+  get endTime(): Date {
+    if (!this.clock) {
+      this.logger.debug('calling endTime() on null clock');
+      return new Date();
+    }
+
+    return this.clock.endTime;
+  }
+
+  /**
+   * Gives a timestap that indicates the span's duration in RFC3339 UTC
+   * "Zulu" format.
+   */
+  get duration(): number {
+    if (!this.clock) {
+      this.logger.debug('calling duration() on null clock');
+      return 0;
+    }
+
+    return this.clock.duration;
+  }
+
+  /** Gives the TraceContext of the span. */
+  get spanContext(): types.SpanContext {
+    return {
       traceId: this.traceId,
       spanId: this.id,
-      name: this.name,
+      options: 0x1,  // always traced
```

Comment:
I'm assuming this was adapted from `SpanBase` - will it cause all spans to be sampled? @mayurkale22

Reply:
Hmm... I didn't touch this line, just moved it.

Reply:
> I'm assuming this was adapted from SpanBase - will it cause all spans to be sampled? 

No. The Sampling Decision happens at the [top level ](https://github.com/census-instrumentation/opencensus-node/blob/master/packages/opencensus-core/src/trace/model/tracer.ts#L150)(when we create the root span using `startRootSpan`). Depending on the decision, it creates either `RootSpan` (which has options set to 0x1) or `NoRecordRootSpan` (options set to 0x0). Does it make sense?